folding-schemes/src/folding/circuits/decider/mod.rs

```diff
@@ -0,0 +1,211 @@
+    }
+
+    pub fn get_challenges_gadget<
+        C: CurveGroup,
+        S: CryptographicSponge,
+        T: TranscriptVar<CF1<C>, S>,
+        U: CommittedInstanceVarOps<C>,
+    >(
+        transcript: &mut T,
+        U_i: &U,
+    ) -> Result<Vec<FpVar<CF1<C>>>, SynthesisError> {
+        let mut challenges = vec![];
+        for cm in U_i.get_commitments() {
+            transcript.absorb(&cm.to_constraint_field()?)?;
+            challenges.push(transcript.get_challenge()?);
+        }
+        Ok(challenges)
+    }
+}
+
+/// Gadget that interpolates the polynomial from the given vector and returns
+/// its evaluation at the given point.
+/// It also offers the rust native implementation compatible with the gadget.
+pub struct EvalGadget {}
+
+impl EvalGadget {
+    pub fn evaluate_native<F: PrimeField>(v: &[F], point: F) -> Result<F, Error> {
+        let mut v = v.to_vec();
+        v.resize(v.len().next_power_of_two(), F::zero());
+
+        Ok(poly_from_vec(v)?.evaluate(&point))
+    }
+
+    pub fn evaluate_gadget<F: PrimeField>(
+        v: &[FpVar<F>],
+        point: &FpVar<F>,
+    ) -> Result<FpVar<F>, SynthesisError> {
+        let mut v = v.to_vec();
+        v.resize(v.len().next_power_of_two(), FpVar::zero());
+        let n = v.len() as u64;
+        let gen = F::get_root_of_unity(n).ok_or(SynthesisError::PolynomialDegreeTooLarge)?;
+        // `unwrap` below is safe because `Radix2DomainVar::new` only fails if
+        // `offset.enforce_not_equal(&FpVar::zero())` returns an error.
+        // But in our case, `offset` is `FpVar::one()`, i.e., both operands of
+        // `enforce_not_equal` are constants.
+        // Consequently, `FpVar`'s implementation of `enforce_not_equal` will
+        // always return `Ok(())`.
+        let domain = Radix2DomainVar::new(gen, log2(v.len()) as u64, FpVar::one()).unwrap();
+
+        let evaluations_var = EvaluationsVar::from_vec_and_domain(v, domain, true);
+        evaluations_var.interpolate_and_evaluate(point)
+    }
+}
+
+/// This is a temporary workaround for step 6 (running NIFS.V for group elements
+/// in circuit) in an NIFS-agnostic way, because different folding schemes have
+/// different interfaces of folding verification now.
+///
+/// In the future, we may introduce a better solution that uses a trait for all
+/// folding schemes that specifies their native and in-circuit behaviors.
+pub trait DeciderEnabledNIFS<
+    C: CurveGroup,
+    RU: CommittedInstanceOps<C>, // Running instance
+    IU: CommittedInstanceOps<C>, // Incoming instance
+    W: WitnessOps<CF1<C>>,
+    A: Arith<W, RU>,
+>
+{
+    type ProofDummyCfg;
+    type Proof: Dummy<Self::ProofDummyCfg>;
+    type RandomnessDummyCfg;
+    type Randomness: Dummy<Self::RandomnessDummyCfg>;
+
+    /// Fold the field elements in `U` and `u` inside the circuit.
+    ///
+    /// `U_vec` is `U` expressed as a vector of `FpVar`s, which can be reused
+    /// before or after calling this function to save constraints.
```

Comment:
@winderica can explain better the specifics in this call, but for context: this comes from [Nova impl](
https://github.com/privacy-scaling-explorations/sonobe/tree/6d8f297f116e1b367d57cc7eb1cf1a5a72eec505/folding-schemes/src/folding/nova/decider_eth_circuit.rs#L399). The idea was that in the [`U.hash()`](
https://github.com/privacy-scaling-explorations/sonobe/tree/6d8f297f116e1b367d57cc7eb1cf1a5a72eec505/folding-schemes/src/folding/traits.rs#L66) call, internally the vector of field elements that represents all the parameters (including points coordinates) of `U` is computed, and later [in other parts of the logic](
https://github.com/privacy-scaling-explorations/sonobe/tree/6d8f297f116e1b367d57cc7eb1cf1a5a72eec505/folding-schemes/src/folding/nova/decider_eth_circuit.rs#L481) of the decider circuit this vector is needed again. So instead of computing it 2 times (inside the circuit), it is computed once and returned so that it can be reused without needing more constraints later.

Reply:
Thanks @arnaucube! The use of `U_vec` here is exactly inspired by `CommittedInstanceVarOps::hash`.

The reason why reusing `U_vec` saves constraints, or why recalculating `U_vec` requires constraints, is due to the implementation of `NonNativeUintVar::to_constraint_field`.
https://github.com/privacy-scaling-explorations/sonobe/blob/6d8f297f116e1b367d57cc7eb1cf1a5a72eec505/folding-schemes/src/folding/circuits/nonnative/uint.rs#L509-L521

In a `NonNativeUintVar`, there are many limbs, and each limb currently stores a `FpVar` that has a small number of bits (much smaller than the field order) for efficiency of integer operations like addition and multiplication. However, a small number of bits per limb will result in more limbs, and consequently, hashing a `NonNativeUintVar` will require more rounds of permutations and thus more constraints.

To save the costs of hashing, we convert the limbs of `NonNativeUintVar` to a compact form in `NonNativeUintVar::to_constraint_field`, such that each new limb now stores as many bits as possible (i.e., `F::MODULUS_BIT_SIZE as usize - 1`), and the number of limbs becomes as small as possible. This process introduces constraints because it decomposes all limbs into bits and then recomposes the bits back to `FpVar`s.

By reusing `U_vec`, we no longer need to call `NonNativeUintVar::to_constraint_field` multiple times, thus saving constraints.

Having said that, I think it is possible to avoid constraints in `NonNativeUintVar::to_constraint_field` while producing short hash preimages. To this end, we do not decompose and recompose limbs. Instead, we directly compose several limbs into one via linear combination, which does not introduce any constraints. Although the number of bits in each new limb is unlikely to be the optimal value `F::MODULUS_BIT_SIZE as usize - 1`, the number of limbs should still be pretty short. In this way, converting `U` to `FpVar`s becomes nearly free, and we may remove `U_vec` in favor of a user friendlier API.